internal/k8s/vcf_controller.go

```diff
@@ -0,0 +1,181 @@
+	NCPBootstrapHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap ADD Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap Update Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+	}
+	c.dynamicInformers.NCPBootstrapInformer.Informer().AddEventHandler(NCPBootstrapHandler)
+
+	go c.dynamicInformers.NCPBootstrapInformer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, c.dynamicInformers.NCPBootstrapInformer.Informer().HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+	} else {
+		utils.AviLog.Info("Caches synced for NCP Bootstrap informer")
+	}
+}
+
+// HandleVCF checks if avi secret used by AKO is already present. If found, the it would try to connect to
+// AVI Controller. If there is any failure, we would look at Bootstrap CR used by NCP to communicate with AKO.
+// If Bootstrap CR is not found, AKO would wait for it to be created. If the authtoken from Bootstrap CR
+// can be used to connect to the AVI Controller, then avi-secret would be created with that token.
+func (c *AviController) HandleVCF(informers K8sinformers, stopCh <-chan struct{}, ctrlCh chan struct{}) {
+	cs := c.informers.ClientSet
+	aviSecret, err := cs.CoreV1().Secrets(utils.GetAKONamespace()).Get(context.TODO(), lib.AviSecret, metav1.GetOptions{})
+	if err == nil {
+		ctrlIP := os.Getenv(utils.ENV_CTRL_IPADDRESS)
+		authToken := aviSecret.Data["authToken"]
+		username := aviSecret.Data["username"]
+		var transport *http.Transport
+		_, err = clients.NewAviClient(
+			ctrlIP, string(username), session.SetAuthToken(string(authToken)),
+			session.SetRefreshAuthTokenCallbackV2(utils.GetAuthtokenFromCache),
+			session.SetNoControllerStatusCheck, session.SetTransport(transport),
+			session.SetInsecure,
+		)
+		if err == nil {
+			utils.AviLog.Infof("Successfully connected to AVI controller using existing AKO secret")
+			return
+		} else {
+			utils.AviLog.Error("AVI controller initialization failed with err: %v", err)
+		}
+	}
+
+	utils.AviLog.Infof("Got error while fetching avi-secret: %v", err)
+
+	if !c.ValidBSData() {
+		utils.AviLog.Infof("Running in a VCF Cluster, but valid Bootstrap CR not found, waiting .. ")
+		startSyncCh := make(chan struct{})
+		c.AddNCPBootstrapEventHandler(informers, stopCh, startSyncCh)
+	L:
+		for {
+			select {
+			case <-startSyncCh:
+				break L
+			case <-ctrlCh:
+				return
+			}
+		}
+	}
+	utils.AviLog.Infof("NCP Bootstrap CR found, continuing AKO initialization")
+	c.UpdateAviSecret()
+}
+
+func (c *AviController) UpdateAviSecret() error {
+	secretName, ns, username := lib.GetBootstrapCRData()
+	cs := c.informers.KubeClientIntf.ClientSet
```

Comment:
is this variable really the clientset or is it the result of the `Get` call?

Reply:
This is clientset, but KubeClientIntf is anonymous field - so ClientSet can be directly accessed, updated.